contrib/epee/include/serialization/wire/read.h

```diff
@@ -0,0 +1,532 @@
+// permitted provided that the following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice, this list of
+//    conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list
+//    of conditions and the following disclaimer in the documentation and/or other
+//    materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be
+//    used to endorse or promote products derived from this software without specific
+//    prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#pragma once
+
+#include <cstdint>
+#include <limits>
+#include <string>
+#include <type_traits>
+
+#include "byte_slice.h"
+#include "serialization/wire/epee/fwd.h"
+#include "serialization/wire/error.h"
+#include "serialization/wire/field.h"
+#include "serialization/wire/fwd.h"
+#include "serialization/wire/traits.h"
+#include "span.h"
+
+/*
+  Custom types (e.g. `type` in namespace `ns`) can define an input function by:
+    * `namespace wire { template<> struct is_blob<ns::type> : std::true_type {}; }`
+    * `namespace wire { void read_bytes(writer&, ns::type&); }`
+    * `namespace ns { void read_bytes(wire::writer&, type&); }`
+
+  See `traits.h` for `is_blob` requirements. `read_bytes` function can also
+  specify derived type for faster output (i.e.
+  `namespace ns { void read_bytes(wire::epee_reader&, type&); }`). Using the
+  derived type allows the compiler to de-virtualize and allows for custom
+  functions not defined by base interface. Using base interface allows for
+  multiple formats with minimal instruction count. */
+
+namespace wire
+{
+  //! Interface for converting "wire" (byte) formats to C/C++ objects without a DOM.
+  class reader
+  {
+    std::size_t depth_; //!< Tracks number of recursive objects and arrays
+
+  protected:
+    epee::span<const std::uint8_t> remaining_; //!< Derived class tracks unprocessed bytes here
+
+    //! \throw wire::exception if max depth is reached
+    void increment_depth();
+    void decrement_depth() noexcept { --depth_; }
```

Comment:
Protected method may cause private field to underflow

Reply:
Iff there's a mismatch in `increment_depth()` calls. Are you advocating for a runtime check here, to guard against that too?

Reply:
I think the best design would be to make `increment_depth` and `decrement_depth` private, add a runtime check to `decrement_depth`, and then have `start_object` and `start_array` both call `increment_depth`, just like how `end_object` and `end_array` call `decrement_depth`.

This is because a `reader` implementation or any buggy serialization code for any type could cause this value to underflow. Also, this value merely acts (from my PoV) as a DDoS prevention measure. I can't think of when the actual value of `depth_` by itself would be useful to a format implementation. For this reason, I think it would be cleaner and more stable to hide the depth value, and have subclasses call the parent `start/end_object/array` methods as appropriate.

Reply:
Also, if there happens to be different structural limits we want to impose (e.g. only on objects) and we can do it in this base class, if we force the subclasses to go thru `start/end_object/array` methods, it will be easier to make those adjustments correctly.

Reply:
I'm leaning towards making this change, but I'll first describe how this is handled in the current design. Its not well documented, a short-coming that needs fixing (or else dropped).

The `from_bytes` template method has the following:
```c++
       R in{std::forward<T>(source)};
       bytes(in, dest);
       in.check_complete();
```
this requires that every reader type have a method called `check_complete()`. This method checks to make sure that the tree is "complete":
```c++
   void json_reader::check_complete() const
   {
     if (depth())
       WIRE_DLOG_THROW(error::rapidjson_e(rapidjson::kParseErrorUnspecificSyntaxError), "Unexpected end");
   }
```
So the potential underflow is intentional - either the next call to `increment_depth()` detects an overflow case or the `check_complete()` detects a non-zero depth.

AFAIK, this achieves the same thing, but with one-less branch per call to `decrement_depth()`. It probably looks better to have the check though, and the extra check is trivial in the design.
---------
I left `increment_depth()` and `decrement_depth()` as `protected` so that "skipping" a field could call them. I _think_ this has never been necessary, as `start_object()` and `end_object()`, etc, have been used instead (so at a minimum I would make the functions `private`.

Reply:
After further thought, I will just add the check. It's possible to call `end_object()` enough times to rollback to `0`, which is unintentional, so a runtime-check is needed. I'll also move them to `private`, as I can't recall using them in the derived classes.

Reply:
I added the checks in the other review, but _might_ revert it. The fallout is that `do_start_object()` and `do_start_array(size_t)` can no longer be de-virtualized.